### PR TITLE
dialects: (builtin) add range verifier for `IntegerAttr`

### PR DESCRIPTION
**Pull request by adutilleul**
Follow-up PR of https://github.com/xdslproject/xdsl/pull/1027 to split the changes properly.

### CODE REVIEW

**Review by superlopuh — APPROVED** [review]
LGTM!

**Review by math-fehr — APPROVED** [review]
Really nice! Thanks!

**Review by webmiche — APPROVED** [review]
Thanks 👍

File: xdsl/dialects/builtin.py [hunk 1]
```diff
@@ -389,6 +389,28 @@ def from_int_and_width(value: int, width: int) -> IntegerAttr[IntegerType]:
...
+                    min_value = -(1 << self.typ.width.data)
+                    max_value = 1 << self.typ.width.data
+                case Signedness.SIGNED:
+                    min_value = -(1 << (self.typ.width.data - 1))
+                    max_value = (1 << (self.typ.width.data - 1)) - 1
+                case Signedness.UNSIGNED:
+                    min_value = 0
+                    max_value = (1 << self.typ.width.data) - 1
+                case _:
+                    assert False, "unreachable"
```
[the remark this answers is not in the document]
Comment (superlopuh) [reply xdsl/dialects/builtin.py:405]: Maybe we can add `Unreachable(Exception)` to utils? I'd like a standard way to do this

[the remark this answers is not in the document]
Comment (adutilleul) [reply xdsl/dialects/builtin.py:405]: What about using `assert_never` (https://typing.readthedocs.io/en/latest/source/unreachable.html) ? It also enforces exhaustiveness.

**Pull request merged**